File: lib/dotsync/utils/manifest.rb
# frozen_string_literal: true

require "json"

module Dotsync
  class Manifest
    MANIFESTS_DIR = "dotsync/manifests"

    # @param dest_dir [String] the mapping's destination directory (absolute)
    # @param key [String] manifest filename key (e.g., "xdg_bin")
    # @param xdg_data_home [String] base path for manifest storage
    def initialize(dest_dir:, key:, xdg_data_home:)
      @dest_dir = dest_dir
      @key = key
      @manifest_path = File.join(xdg_data_home, MANIFESTS_DIR, "#{key}.json")
    end

    # Returns array of relative file paths from stored manifest
    # @return [Array<String>]
    def read
      return [] unless File.exist?(@manifest_path)

      data = JSON.parse(File.read(@manifest_path))
      Array(data["files"])
    rescue JSON::ParserError
      []
    end

    # Writes current file list to manifest
    # @param files [Array<String>] relative file paths
    def write(files)
      FileUtils.mkdir_p(File.dirname(@manifest_path))
      File.write(@manifest_path, JSON.pretty_generate({ "files" => files.sort }))
    end

    # Returns orphan absolute paths: files in old manifest but not in current_files
    # @param current_files [Array<String>] relative file paths currently synced
    # @return [Array<String>] absolute paths ready for deletion
    def orphans(current_files)
      previous = read
      orphaned = previous - current_files
      orphaned.map { |file| File.join(@dest_dir, file) }
    end
  end
end

File: internal/manifest/manifest.go
// Package manifest tracks which files an `only`-filtered mapping previously
// synced, so orphans (files dropped from the filter) can be removed on pull
// without force mode. Ports Dotsync::Manifest (a JSON file per shorthand key).
package manifest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"sort"
)

const manifestsDir = "dotsync/manifests"

// Manifest is the record for one mapping's destination, keyed by manifest key.
type Manifest struct {
	destDir      string
	manifestPath string
}

// New builds a Manifest at $xdgDataHome/dotsync/manifests/<key>.json.
func New(destDir, key, xdgDataHome string) *Manifest {
	return &Manifest{
		destDir:      destDir,
		manifestPath: filepath.Join(xdgDataHome, manifestsDir, key+".json"),
	}
}

type doc struct {
	Files []string `json:"files"`
}

// Read returns the relative file paths recorded in the manifest, or nil.
func (m *Manifest) Read() []string {
	data, err := os.ReadFile(m.manifestPath)
	if err != nil {
		return nil
	}
	var d doc
	if json.Unmarshal(data, &d) != nil {
		return nil
	}
	return d.Files
}

// Write records the current relative file list (sorted).
func (m *Manifest) Write(files []string) error {
	sorted := append([]string(nil), files...)
	sort.Strings(sorted)
	if err := os.MkdirAll(filepath.Dir(m.manifestPath), 0o755); err != nil {
		return err
	}
	b, err := json.MarshalIndent(doc{Files: sorted}, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(m.manifestPath, b, 0o644)
}

// Orphans returns absolute paths that were recorded previously but are no longer
// in currentFiles.
func (m *Manifest) Orphans(currentFiles []string) []string {
	current := make(map[string]bool, len(currentFiles))
	for _, f := range currentFiles {
		current[f] = true
	}
	var orphans []string
	for _, f := range m.Read() {
		if !current[f] {
			orphans = append(orphans, filepath.Join(m.destDir, f))
		}
	}
	return orphans
}
